src/script/solver.h

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2009-2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_SCRIPT_SOLVER_H
+#define BITCOIN_SCRIPT_SOLVER_H
+
+#include <attributes.h>
+#include <script/script.h>
+
+#include <string>
+#include <optional>
+#include <utility>
+#include <vector>
+
+class CPubKey;
+template <typename C> class Span;
+
+enum class TxoutType {
+    NONSTANDARD,
+    // 'standard' transaction types:
+    PUBKEY,
+    PUBKEYHASH,
+    SCRIPTHASH,
+    MULTISIG,
+    NULL_DATA, //!< unspendable OP_RETURN script that carries data
+    WITNESS_V0_SCRIPTHASH,
+    WITNESS_V0_KEYHASH,
+    WITNESS_V1_TAPROOT,
+    WITNESS_UNKNOWN, //!< Only for Witness versions not already defined above
+};
+
+/** Get the name of a TxoutType as a string */
+std::string GetTxnOutputType(TxoutType t);
+
+constexpr bool IsPushdataOp(opcodetype opcode)
+{
+    return opcode > OP_FALSE && opcode <= OP_PUSHDATA4;
+}
+
+/**
+ * Parse a scriptPubKey and identify script type for standard scripts. If
+ * successful, returns script type and parsed pubkeys or hashes, depending on
+ * the type. For example, for a P2SH script, vSolutionsRet will contain the
+ * script hash, for P2PKH it will contain the key hash, etc.
+ *
+ * @param[in]   scriptPubKey   Script to parse
+ * @param[out]  vSolutionsRet  Vector of parsed pubkeys and hashes
+ * @return                     The script type. TxoutType::NONSTANDARD represents a failed solve.
```

Comment:
It could be noted what represents a “failed solve” is up to the caller. Actually `TxoutType::WITNESS_UNKNOWN` is considered as a failure in `AreInputsStandard`, as witness unknown is a special case of non-standardness for undefined segwit output.

Reply:
leaving as is since it is only moved.